index.js

```diff
@@ -10,14 +11,12 @@ function encoder (encryptionKey, opts = {}) {
   assert(Buffer.isBuffer(encryptionKey), 'encryption key must be a buffer')
   assert(encryptionKey.length === sodium.crypto_secretbox_KEYBYTES, `cobox-crypto: key must be a buffer of length ${sodium.crypto_secretbox_KEYBYTES}`)
 
-  opts.valueEncoding = _resolveStringEncoder(opts.valueEncoding)
+  const encoder = codecs(opts.valueEncoding)
 
   return {
     encode (message, buffer, offset) {
       // Run originally provided encoder if any
-      if (opts.valueEncoding && typeof opts.valueEncoding.encode === 'function') {
-        message = opts.valueEncoding.encode(message, buffer, offset)
-      }
+      if (opts.valueEncoding) message = encoder.encode(message, buffer, offset)
       if (!Buffer.isBuffer(message)) message = Buffer.from(message, 'utf-8')
```

Comment:
I believe this line to be dangerous: if the input is i.e. a string it will turn it into a buffer on encoding but on decoding it will be returned as buffer and not as string. I believe it is safer to use the `encoder` as is - always both while encoding and decoding for a consistent behavior.

Reply:
can you explain a bit more @martinheidegger? Im not sure if you mean the previous way, or using codecs. Codecs should pick the correct original encoder, as long as it was passed in correctly in the first place - so maybe I'm missing something?

Reply:
`encoder.encode()` will always return a Buffer _(if it implements the interface properly)_, which means that `message` in the next line will only not be a Buffer if `opts.valueEncoding` is not given. `const encoder = codecs(opts.valueEncoding)` on the other hand will always contain an encoder.

I would turn it into:

```javascript
    encode (message, buffer, offset) {
      // Run originally provided encoder if any
      message = encoder.encode(message, buffer, offset)
      const ciphertext = Buffer.alloc(message.length + sodium.crypto_secretbox_MACBYTES)
      const nonce = Buffer.alloc(sodium.crypto_secretbox_NONCEBYTES)
      sodium.randombytes_buf(nonce)
      sodium.crypto_secretbox_easy(ciphertext, message, nonce, encryptionKey)
      zero(message)
      return Buffer.concat([nonce, ciphertext])
    },
    decode (buffer, start, end) {
      const nonce = buffer.slice(0, sodium.crypto_secretbox_NONCEBYTES)
      const messageWithMAC = buffer.slice(sodium.crypto_secretbox_NONCEBYTES)
      const message = Buffer.alloc(messageWithMAC.length - sodium.crypto_secretbox_MACBYTES)
      assert(
        sodium.crypto_secretbox_open_easy(message, messageWithMAC, nonce, encryptionKey),
        'Decryption failed!'
      )
      // Run originally provided encoder if any
      return encoder.decode(message, start, end)
    }
```

Reply:
nice one martin, I've changed to suit this case.